### PR TITLE
Make ast_utils into a library

### PR DESCRIPTION
**Pull request by dschuff**
Split ExpressionAnalyzer and ExpressionManipulator into cpp files, and turn their giant template functions into simple functions which take a callback.
More organization, fewer mammoth headers, makes the build a few seconds faster, and the binaries a couple MB smaller.

### CODE REVIEW

File: src/ast/ExpressionManipulator.cpp [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+  } else {
+    // we need to make room
+    list.push_back(nullptr);
+    for (Index i = list.size() - 1; i > index; i--) {
+      list[i] = list[i - 1];
+    }
+    list[index] = add;
+  }
+}
+}
```
Comment (jgravelle-google) [inline src/ast/ExpressionManipulator.cpp]: Should probably be consistent with the trailing `} // namespace wasm`

File: src/ast_utils.h [hunk 2]
```diff
@@ -457,431 +282,18 @@ struct ExpressionAnalyzer {
...
-        }
-        default: WASM_UNREACHABLE();
-      }
-      #undef CHECK
-      #undef PUSH
-    }
-    if (leftStack.size() > 0 || rightStack.size() > 0) return false;
-    return true;
-  }
+  using ExprComparer = std::function<bool(Expression*, Expression*)>;
```
Comment (kripken) [inline src/ast_utils.h]: what is `using`?

Comment (dschuff) [inline src/ast_utils.h]: It's basically a nicer way of spelling `typedef` in C++11: http://en.cppreference.com/w/cpp/language/type_alias

File: src/passes/OptimizeInstructions.cpp [hunk 3]
```diff
@@ -103,10 +103,7 @@ struct Match {
   bool check(Expression* seen) {
     // compare seen to the pattern input, doing a special operation for our "wildcards"
     assert(wildcards.size() == 0);
-    return ExpressionAnalyzer::flexibleEqual(pattern.input, seen, *this);
-  }
-
-  bool compare(Expression* subInput, Expression* subSeen) {
+    auto compare = [this](Expression* subInput, Expression* subSeen) {
     CallImport* call = subInput->dynCast<CallImport>();
```
Comment (kripken) [inline src/passes/OptimizeInstructions.cpp]: indentation looks wrong

File: src/passes/OptimizeInstructions.cpp [hunk 4]
```diff
@@ -137,18 +134,19 @@ struct Match {
...
-    return ExpressionManipulator::flexibleCopy(pattern.output, wasm, *this);
-  }
-
-  // When copying a wildcard, perform the substitution.
-  // TODO: we can reuse nodes, not copying a wildcard when it appears just once, and we can reuse other individual nodes when they are discarded anyhow.
-  Expression* copy(Expression* curr) {
+    // When copying a wildcard, perform the substitution.
+    // TODO: we can reuse nodes, not copying a wildcard when it appears just once, and we can reuse other individual nodes when they are discarded anyhow.
+    auto copy = [this](Expression* curr) -> Expression* {
     CallImport* call = curr->dynCast<CallImport>();
```
Comment (kripken) [inline src/passes/OptimizeInstructions.cpp]: ditto about indentation

**Comment by kripken** [comment]
Overall looks like a nice improvement, thanks.

In addition to the comments above, we should verify performance does not regress - the change to use a lambda function is on some hot methods.

File: src/passes/OptimizeInstructions.cpp (hunk 3 above)
Comment (dschuff) [inline src/passes/OptimizeInstructions.cpp]: Fixed

File: src/passes/OptimizeInstructions.cpp (hunk 4 above)
Comment (dschuff) [inline src/passes/OptimizeInstructions.cpp]: Fixed

**Comment by dschuff** [comment]
Sure; just run `wasm-opt` over a large input, or do you have a better benchmark?

File: src/ast_utils.h (hunk 2 above)
Comment (kripken) [inline src/ast_utils.h]: cool, never seen it before

**Comment by kripken** [comment]
Yeah, `wasm-opt -O` on a not-yet optimized large input is good enough.

**Comment by dschuff** [comment]
Biggest wasm I have handy is poppler (are any of the emscripten benchmarks or other tests bigger than ~2M)? and any differences seem in the noise on my machine.

**Comment by kripken** [comment]
Ok. Hmm, we had bigger builds in the build-suite, but they are not updated. Maybe I should test on a Unity build, I have one locally. Is now a good time to check or are changes still landing in this branch?

**Comment by dschuff** [comment]
No, this branch should be ready to go now.

**Comment by kripken** [comment]
Ok, tested on Unity, results look fine. lgtm.

**Pull request merged**